synapse/storage/databases/main/devices.py

```diff
@@ -1836,7 +1836,7 @@ def get_uncoverted_outbound_room_pokes_txn(
                     device_id,
                     room_id,
                     stream_id,
-                    db_to_json(opentracing_context),
```

Comment:
I wasn't sure I liked this one. The original complaint was

> synapse/storage/databases/main/devices.py:1860: error: Argument 2 to "runInteraction" of "DatabasePool" has incompatible type "Callable[[LoggingTransaction], List[Tuple[str, str, str, int, Optional[Dict[str, str]]]]]"; expected "Callable[..., List[Tuple[str, str, str, int, Dict[str, str]]]]"  [arg-type]

Which corresponds to the `context` argument of this function:

https://github.com/matrix-org/synapse/blob/9dc3293e0b3a5cbf6fcc4a0cef7386b531190882/synapse/storage/databases/main/devices.py#L1709-L1756

I don't think we want to be encoding `context=None` to a JSON `null` here!

Reply:
Note that `opentracing_context` is a nullable text field in the DB:

```
matrix=> \d device_lists_changes_in_room
Did not find any relation named "device_lists_changes_in_room".
matrix=> \d matrix.device_lists_changes_in_room
             Table "matrix.device_lists_changes_in_room"
          Column           |  Type   | Collation | Nullable | Default 
---------------------------+---------+-----------+----------+---------
 user_id                   | text    |           | not null | 
 device_id                 | text    |           | not null | 
 room_id                   | text    |           | not null | 
 stream_id                 | bigint  |           | not null | 
 converted_to_destinations | boolean |           | not null | 
 opentracing_context       | text    |           |          | 
Indexes:
    "device_lists_changes_in_stream_id" UNIQUE, btree (stream_id, room_id)
    "device_lists_changes_in_stream_id_unconverted" btree (stream_id) WHERE NOT converted_to_destinations
```

Reply:
Do you know if we have any of that sort of data here though? It is very unclear to me if it is safe to change this.

Reply:
Not on matrix.org:

```
matrix=> select * from matrix.device_lists_changes_in_room where opentracing_context='null' limit 1;
(0 rows)

Time: 11322.423 ms (00:11.322)
matrix=> select * from matrix.device_lists_changes_in_room where opentracing_context='"null"' limit 1;
(0 rows)

Time: 10651.809 ms (00:10.652)
matrix=> select * from matrix.device_lists_changes_in_room where opentracing_context is NULL limit 1;
(0 rows)

Time: 3886.793 ms (00:03.887)
```

I suppose a safer approach would be to write a migration which makes the column non-nullable.

Maybe this is better dropped though; ideally type hints wouldn't kick off invasive changes.

Reply:
Thanks. I'd be curious what @erikjohnston thinks here since this was was just added in #12321 (and maybe #13045). Is there a long term plan here we're unsure about?

Reply:
I'm a bit confused as to why we can't let it be `None` if I'm honest?

Reply:
I was mainly just trying to keep the existing type hints happy. Some of them say `Dict[]` and some of them say `Optional[Dict]`.

Maybe we just use `Optional[Dict]` everywhere?

Reply:
Yeah, I'd go with `Optional[Dict]` if the DB has a nullable column?

Reply:
8395995e0 and 60e2ce018 do this.